src/system.rs

```diff
@@ -13,6 +14,174 @@ use std::convert::TryInto;
 use std::io::ErrorKind;
 use std::rc::Rc;
 
+pub type SharedContext = Rc<RefCell<Context>>;
+
+/// Defines a [Context] struct which stores information on parent types
+/// while loading a [Schema] with types which is used to resolve self referencing schema types
+#[derive(Debug, Clone)]
+pub struct Context {
+    ids_by_name: HashMap<String, TypeId>,
+    parent: Option<(String, TypeId)>,
+    types_by_id: Vec<Option<TypeDefinition>>, // a None in this vector represents unresolved type
+}
+
+impl Context {
+    pub fn new() -> Self {
+        Self {
+            parent: None,
+            ids_by_name: HashMap::new(),
+            types_by_id: Vec::new(),
+        }
+    }
+
+    /// Adds all the types from Context into given TypeStore and clears Context types for loading next set of types
+    /// this is used after a schema named type/root type is loaded entirely into Context
+    pub fn update_type_store(&mut self, type_store: &SharedTypeStore) -> IonSchemaResult<()> {
+        for optional_type in &self.types_by_id {
+            // return an error if any of the type in types_by_id vector is None/Unresolved
+            let type_def = optional_type
+                .to_owned()
+                .ok_or(unresolvable_schema_error_raw(
+                    "Unable to load schema due to unresolvable type",
+                ))?;
+            match type_def.to_owned().name() {
+                Some(name) => type_store.borrow_mut().add_named_type(name, type_def),
+                None => type_store.borrow_mut().add_anonymous_type(type_def),
+            };
+        }
+        self.types_by_id.clear();
+        self.ids_by_name.clear();
+        Ok(())
+    }
+
+    /// Returns total number of types stored in the [TypeStore]
+    pub fn get_total_types(&self) -> usize {
+        self.types_by_id.len()
+    }
+
+    /// Provides the [TypeDefinition] associated with given name if it exists in the [TypeStore]  
+    /// Otherwise returns None
+    pub fn get_type_by_name(
+        &self,
+        name: &str,
+        type_store: &SharedTypeStore,
+    ) -> Option<TypeDefinition> {
+        self.ids_by_name
+            .get(name)
+            .and_then(|id| match self.types_by_id.get(*id) {
+                Some(t) => Some(t.to_owned().unwrap()),
+                None => match type_store.borrow().get_type_by_id(id.to_owned()) {
+                    Some(t) => Some(t.to_owned()),
+                    None => None,
+                },
+            })
+    }
+
+    /// Provides the [TypeId] associated with given name if it exists in the [TypeStore] or [Context]  
+    /// Otherwise returns None
+    pub fn get_type_id_by_name(&self, name: &str, type_store: &SharedTypeStore) -> Option<TypeId> {
+        match self.ids_by_name.get(name) {
+            Some(id) => Some(id.to_owned()),
```

Comment:
There are a lot of `to_owned()`s in this file that I think are unnecessary? `TypeId` is just a `usize`, so it can be freely copied everywhere.